src/Bicep.Decompiler/Rewriters/ForExpressionSimplifierRewriter.cs

```diff
@@ -0,0 +1,190 @@
+    {
+        private readonly SemanticModel semanticModel;
+
+        public ForExpressionSimplifierRewriter(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        private static bool IsLoopIteratorExpression(SyntaxBase syntax, [NotNullWhen(true)] out VariableAccessSyntax? arrayVariable)
+        {
+            arrayVariable = null;
+
+            if (syntax is not FunctionCallSyntax rangeFunction ||
+                !LanguageConstants.IdentifierComparer.Equals(rangeFunction.Name.IdentifierName, "range") ||
+                rangeFunction.Arguments.Length != 2)
+            {
+                return false;
+            }
+
+            if (rangeFunction.Arguments[0].Expression is not IntegerLiteralSyntax startRange ||
+                startRange.Value != 0)
+            {
+                return false;
+            }
+
+            if (rangeFunction.Arguments[1].Expression is not FunctionCallSyntax lengthFunction ||
+                !LanguageConstants.IdentifierComparer.Equals(lengthFunction.Name.IdentifierName, "length") ||
+                lengthFunction.Arguments.Length != 1)
+            {
+                return false;
+            }
+
+            if (lengthFunction.Arguments[0].Expression is not VariableAccessSyntax variableAccess)
+            {
+                return false;
+            }
+
+            arrayVariable = variableAccess;
+            return true;
+        }
+
+        protected override SyntaxBase ReplaceForSyntax(ForSyntax syntax)
+        {
+            syntax = (ForSyntax)base.ReplaceForSyntax(syntax);
+
+            if (!IsLoopIteratorExpression(syntax.Expression, out var arrayVariable))
+            {
+                return syntax;
+            }
+
+            if (syntax.VariableSection is not LocalVariableSyntax indexVariable)
+            {
+                return syntax;
+            }
+
+            var arraySymbol = semanticModel.GetSymbolInfo(arrayVariable);
+            var arrayIndexSymbol = semanticModel.GetSymbolInfo(indexVariable);
+
+            if (arraySymbol is null || arrayIndexSymbol is null)
+            {
+                return syntax;
+            }
+
+            var arrayAccesses = new HashSet<ArrayAccessSyntax>();
+            var independentIndexAccesses = new HashSet<VariableAccessSyntax>();
+            CallbackVisitor.Visit(syntax, child => {
+                if (child is ArrayAccessSyntax arrayAccess)
+                {
+                    if (semanticModel.GetSymbolInfo(arrayAccess.BaseExpression) == arraySymbol && 
+                        semanticModel.GetSymbolInfo(arrayAccess.IndexExpression) == arrayIndexSymbol)
+                    {
+                        arrayAccesses.Add(arrayAccess);
+
+                        // we don't want to count the VariableAccessSyntax under this particular node,
+                        // so return false to skip visiting children.
+                        return false;
```

Comment:
>false [](start = 31, length = 5)

If the `BaseExpression` is a `VariableAccessSyntax` and IndexExpression` is also a `VariableAccessSyntax` then we're basically skipping visits of the child tokens pretty much? (No issue - just double checking my own understanding of the logic.)

Reply:
Yeah, exactly. Basically what I'm trying to check is "are there any accesses to `i` outside of `myArray[i]`", so I'm trying to avoid visiting the `i` inside the array access.

This is necessary to decide between generating `[for thing in ...` vs `[for (thing, i) in ...` (e.g. "Do I still need to provide access to `i` in the rewritten access?").